onnxruntime/core/mlas/lib/qgemm.cpp

```diff
@@ -84,120 +82,240 @@ MlasGemmU8X8ScaleSumBuffer(
+        }
+    }
+}
+
+template<bool HasBias, MLAS_QGEMM_OUTPUT_MODE Mode, MLAS_QUANTIZATION_GRANULARITY QuantGran>
+inline
 void
-MlasGemmU8X8OutputFloat(
-    const MLAS_GEMM_U8X8_WORK_BLOCK* WorkBlock,
-    int32_t* C,
+MLAS_QGEMM_SCALE_BIAS_OUTPUT_PROCESSOR::ProcessImpl(
+    const int32_t* C,
+    size_t StartM,
     size_t StartN,
     size_t CountM,
-    size_t CountN
-    )
+    size_t CountN,
+    size_t ldc) const
 /*++
 
 Routine Description:
 
-    This routine converts the output matrix to a floating point format using
-    the supplied scale and bias parameters.
+    This routine converts the output matrix C to a floating point format using
+    the stored scale and bias parameters.
 
 Arguments:
 
-    WorkBlock - Supplies the structure containing the GEMM parameters.
-
     C - Supplies the address of matrix C.
 
-    StartN - Supplies the starting column offset relative to the base of the
-        work block. This is used to offset into column vectors accessed via the
-        work block.
+    StartM - Supplies the starting row offset relative to the matrix.
+
+    StartN - Supplies the starting column offset relative to the matrix.
 
     CountM - Supplies the number of rows of the output matrix to process.
 
     CountN - Supplies the number of columns of the output matrix to process.
 
+    ldc - Supplies the leading dimension of C.
+
 Return Value:
 
     None.
 
 --*/
 {
-    const size_t ldc = WorkBlock->ldc;
-    const MLAS_FLOAT32X4 ScaleVector = MlasBroadcastFloat32x4(WorkBlock->Scale);
+    float* Output = Output_;
+    const float* Bias = Bias_;
+    const float* Scale = Scale_;
+
+    if (HasBias) {
+        Bias += StartN;
+    }
+
+    if(QuantGran == MLAS_QUANTIZATION_GRANULARITY::PerColumn){
+        Scale += StartN;
+    }
+
+    const MLAS_FLOAT32X4 ScaleVector = MlasBroadcastFloat32x4(Scale_);
 #if !defined(MLAS_SSE2_INTRINSICS)
     const float ScaleValue = MlasExtractLaneFloat32x4<0>(ScaleVector);
 #endif
 
-    //
-    // Check if the optional bias vector was supplied.
-    //
+    C += StartM * ldc + StartN;
+    Output += StartM * LeadingDimensionOutput_ + StartN;
+
 
-    const float* BiasFloat = WorkBlock->BiasFloat;
+    while (CountM-- > 0) {
 
-    if (BiasFloat != nullptr) {
+        float* c_out = Output;
+        const int32_t* c = C;
+        const float* bias = Bias;
+        const float* scale = Scale;
 
-        BiasFloat += WorkBlock->RangeStartN + StartN;
+        size_t n = CountN;
 
-        while (CountM-- > 0) {
+        while (n >= 4) {
 
-            const float* bias = BiasFloat;
-            int32_t* c = C;
-            size_t n = CountN;
+            MLAS_FLOAT32X4 FloatVector = MlasCastToFloat32x4(MlasLoadInt32x4(c));
 
-            while (n >= 4) {
+            if (QuantGran == MLAS_QUANTIZATION_GRANULARITY::PerColumn) {
 
-                MLAS_FLOAT32X4 FloatVector = MlasCastToFloat32x4(MlasLoadInt32x4(c));
+                if (Mode == MLAS_QGEMM_OUTPUT_MODE::AccumulateMode) {
```

Comment:
Seems like the code could be refactored to first do the QuantGran stuff to either load a scale vector or use the broadcast value, then have a different if block that does the Mode stuff. There's some duplication of logic here that doesn't seem necessary.